File: 백준/Gold/12744. 팬케이크 쌓기/팬케이크 쌓기.py
# 팬케이크 쌓기
import sys
input = sys.stdin.readline
from collections import deque

N = int(input())
start = []
sett = set()
target = tuple([i for i in range(1, N+1)])

for i in range(N):
    a, b = input().rstrip().split()

    if b == "+":
        start.append(int(a))
    else:
        start.append(-int(a))

que = deque([])
que.append((tuple(start), 0))

while que:
    cake, cnt = que.popleft()

    if cake == target:
        break

    if cake in sett:
        continue
    else:
        sett.add(cake)

    for i in range(N):
        tmp = list(cake)

        for j in range(i//2+1):
            tmp[j], tmp[i-j] = -tmp[i-j], -tmp[j]

        tmp = tuple(tmp)
        
        if tmp in sett:
            continue
        else:
            que.append((tmp, cnt+1))

print(cnt)

File: 백준/Gold/12744. 팬케이크 쌓기/팬케이크 쌓기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N;
    static int[] target;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());
        int[] start = new int[N];
        target = new int[N];
        for (int i = 0; i < N; i++) {
            String[] input = br.readLine().split(" ");
            int num = Integer.parseInt(input[0]);
            if (input[1].equals("+")) {
                start[i] = num;
            } else {
                start[i] = -num;
            }
            target[i] = i + 1;
        }

        Queue<State> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        queue.add(new State(start, 0));

        while (!queue.isEmpty()) {
            State s = queue.poll();
            int[] cake = s.arr;
            int cnt = s.cnt;

            if (Arrays.equals(cake, target)) {
                System.out.println(cnt);
                return;
            }

            String key = Arrays.toString(cake);
            if (visited.contains(key)) continue;
            visited.add(key);

            for (int i = 0; i < N; i++) {
                int[] tmp = Arrays.copyOf(cake, N);
                for (int j = 0; j <= i / 2; j++) {
                    int t = tmp[j];
                    tmp[j] = -tmp[i - j];
                    tmp[i - j] = -t;
                }

                String newKey = Arrays.toString(tmp);
                if (!visited.contains(newKey)) {
                    queue.add(new State(tmp, cnt + 1));
                }
            }
        }
    }

    static class State {
        int[] arr;
        int cnt;

        State(int[] arr, int cnt) {
            this.arr = arr;
            this.cnt = cnt;
        }
    }
}
